substrate/frame/parameters/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+//! performance with convenience and should therefore only be used in places where that is proven to
+//! be uncritical.
+//!
+//! ### Example Configuration
+//!
+//! Here is an example of how to define some parameters, including their default values:
+#![doc = docify::embed!("src/tests/mock.rs", dynamic_params)]
+//!
+//! A permissioned origin can be define on a per-key basis like this:
+#![doc = docify::embed!("src/tests/mock.rs", custom_origin)]
+//!
+//! Now the aggregated parameter needs to be injected into the pallet config:
+#![doc = docify::embed!("src/tests/mock.rs", impl_config)]
+//!
+//! As last step, the parameters can now be used in other pallets 🙌
+#![doc = docify::embed!("src/tests/mock.rs", usage)]
+//!
+//! ### Examples Usage
+//!
+//! Now to demonstrate how the values can be updated:
+#![doc = docify::embed!("src/tests/unit.rs", set_parameters_example)]
+//!
+//! ## Low Level / Implementation Details
+//!
+//! The pallet stores the parameters in a storage map and implements the matching `Get<Value>` for
+//! each `Key` type. The `Get` then accesses the `Parameters` map to retrieve the value. An event is
+//! emitted every time that a value was updated. It is even emitted when the value is changed to the
+//! same.
+//!
+//! The key and value types themselves are defined by macros and aggregated into a runtime wide
+//! enum. This enum is then injected into the pallet. This allows it to be used without any changed
+//! to the pallet that the parameter will be utilized by.
+//!
+//! ### Design Goals
+//!
+//! 1. Easy to update without runtime upgrade.
+//! 2. Exposes metadata and docs for user convenience.
+//! 3. Can be permissioned on a per-key base.
+//!
+//! ### Design
+//!
+//! 1. Everything is done at runtime without the need for `const` values. `Get` allows for this -
+//! which is coincidentally an upside and a downside. 2. The types are defined through macros, which
+//! allows to expose metadata and docs. 3. Access control is done through the `EnsureOriginWithArg`
+//! trait, that allows to pass data along to the origin check. It gets passed in the key. The
+//! implementor can then match on the key and the origin to decide whether the origin is
+//! permissioned to set the value.
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+use frame_support::traits::{
+	dynamic_params::{AggregratedKeyValue, IntoKey, Key, RuntimeParameterStore, TryIntoKey},
+	EnsureOriginWithArg,
+};
+
+mod benchmarking;
+#[cfg(test)]
+mod tests;
+mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// The key type of a parameter.
+type KeyOf<T> = <<T as Config>::RuntimeParameters as AggregratedKeyValue>::AggregratedKey;
+
+/// The value type of a parameter.
+type ValueOf<T> = <<T as Config>::RuntimeParameters as AggregratedKeyValue>::AggregratedValue;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config(with_default)]
+	pub trait Config: frame_system::Config {
```

Comment:
Don't want to bug you much further but I think it is pretty low effort to also provide some `DefaultConfig` impl here, but a follow-up is a part of #171 works I guess.